### PR TITLE
ACS-1404 : Trashcan Cleaner job doesn't work when trying to delete from a big amount of data

### PR DESCRIPTION
**Pull request by DenisGabriela**
- upgrade super pom, alfresco-repository and postgresql
   - add maven profile to raise a postgres container for tests
   - get batch to delete based on deleteBatchCount and keepPeriod properties
   - fix olderThanDaysToKeep() method
   - code cleanup
   - add test for keepPeriod property usage

### CODE REVIEW

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java [hunk 1]
```diff
@@ -93,110 +94,87 @@ public TrashcanCleaner(NodeService nodeService, TransactionService transactionSe
...
+        AuthenticationUtil.runAsSystem(() ->
         {
-            nodeService.deleteNode(nodes.get(i - 1));
-        }
+            RetryingTransactionCallback<Void> txnWork = () ->
+            {
+                nodes.forEach(nodeService::deleteNode);
+                return null;
+            };
+            return transactionService.getRetryingTransactionHelper().doInTransaction(txnWork);
```
Comment (cristianturlica) [inline src/main/java/org/alfresco/trashcan/TrashcanCleaner.java:132]: This calls `doInTransaction(cb, false, false)` and this does not open a new transaction... you need to  call doInTransaction(txnWork, false, true).

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java [hunk 2]
```diff
@@ -93,110 +94,87 @@ public TrashcanCleaner(NodeService nodeService, TransactionService transactionSe
...
     private void deleteNodes(List<NodeRef> nodes)
     {
-        for (int i = nodes.size(); i > 0; i--) 
+        AuthenticationUtil.runAsSystem(() ->
         {
-            nodeService.deleteNode(nodes.get(i - 1));
-        }
+            RetryingTransactionCallback<Void> txnWork = () ->
+            {
+                nodes.forEach(nodeService::deleteNode);
```
Comment (cristianturlica) [inline src/main/java/org/alfresco/trashcan/TrashcanCleaner.java:129]: In the initial version of the code we used descending order to iterate over the nodes. Why do we need to change this?

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java [hunk 3]
```diff
@@ -93,110 +90,71 @@ public TrashcanCleaner(NodeService nodeService, TransactionService transactionSe
...
-    private List<ChildAssociationRef> filterArchiveUsers(List<ChildAssociationRef> allChilds)
-    {
-        List<ChildAssociationRef> children = new ArrayList<>();
-        for (ChildAssociationRef childAssoc : allChilds) 
-        {
-            NodeRef child = childAssoc.getChildRef();
-            if (!ContentModel.TYPE_ARCHIVE_USER.equals(nodeService.getType(child)))
-            {
-                children.add(childAssoc);
-            }
```
Comment (cristianturlica) [inline src/main/java/org/alfresco/trashcan/TrashcanCleaner.java:188]: Why did we remove this filtering?

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java [hunk 4]
```diff
@@ -238,37 +200,24 @@ public void clean()
...
-                    }
-                };
-                return transactionService.getRetryingTransactionHelper().doInTransaction(txnWork);
-            }
+                return null;
+            };
+            return transactionService.getRetryingTransactionHelper().doInTransaction(txnWork, true, true);
         });
 
+        deleteNodes(trashcanNodes);
```
Comment (cristianturlica) [inline src/main/java/org/alfresco/trashcan/TrashcanCleaner.java:219]: Why did we move this outside the transaction?

**Review by skopf — APPROVED** [review]
This is quite a clever solution. It took me some time to actually understand that it is working and why it is working.

In might be helpful to add some more comments to the code to help other engineers working in this area to better understand how this works.

First of all, you are right to only look at the association type instead of filtering by node type.
The next clever bit is to use the batch size as maxItems parameter of the query. But to understand that this actually works, you should explain in a comment that the cm:children association type is defined as an ordered association, and that newly deleted nodes will be added to the end of this list. Since this method in the NodeService is returning the nodes in the correct order, you will always look at the batchSize nodes in the recycle bin that are in there for the longest time.

Also deleting nodes in individual transactions will help to keep transactions smaller and make this more effective.

This is a massive improvement over what we have so far. There could still be the case where we have a large node tree in the trashcan, so that the nodeService.deleteNode(n) invocation is taking a long time. But this is a general problem with the deleteNode() implementation and not necessarily a problem in the trashcan cleaner.

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java (hunk 3 above)
Comment (DenisGabriela) [reply src/main/java/org/alfresco/trashcan/TrashcanCleaner.java:188]: We filter the nodes by association type [here](https://github.com/Alfresco/alfresco-trashcan-cleaner-module/pull/7/files#diff-481b1ed71300f831eb4955aa52c43d3a31d7a51dcd322695c23709523fa285e4R154). This way we get only the oldest _deleteBatchCount_ nodes in trashcan of type ContentModel.ASSOC_CHILDREN.

File: src/main/java/org/alfresco/trashcan/TrashcanCleaner.java (hunk 2 above)
Comment (DenisGabriela) [reply src/main/java/org/alfresco/trashcan/TrashcanCleaner.java]: We don't need to, but it is a bit cleaner.
By testing both ways of iterating over the nodes, I found no importance in the order.

**Comment by DenisGabriela** [comment]
@cristianturlica, @skopf  Thanks for your reviews. I addressed the comments and if there's nothing else, I'll merge the pull request.

**Review by cristianturlica — COMMENTED** [review]
LGTM

**Pull request merged**